core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -418,6 +425,64 @@ public final SegmentsStats segmentsStats() {
                 stats.addTermVectorsMemoryInBytes(guardedRamBytesUsed(segmentReader.getTermVectorsReader()));
                 stats.addNormsMemoryInBytes(guardedRamBytesUsed(segmentReader.getNormsReader()));
                 stats.addDocValuesMemoryInBytes(guardedRamBytesUsed(segmentReader.getDocValuesReader()));
+
+                if (!detailed) {
+                    // Gathering index file stats may be costly. Don't do it unless specified
+                    continue;
+                }
+
+                Directory directory = null;
+                SegmentCommitInfo segmentCommitInfo = segmentReader.getSegmentInfo();
+                boolean useCompoundFile = segmentCommitInfo.info.getUseCompoundFile();
+                if (useCompoundFile) {
+                    try {
+                        directory = engineConfig.getCodec().compoundFormat().getCompoundReader(segmentReader.directory(), segmentCommitInfo.info, IOContext.READ);
+                    } catch (IOException e) {
+                        logger.warn("Error when opening compound reader for Directory [{}] and SegmentCommitInfo [{}]", e,
+                                    segmentReader.directory(), segmentCommitInfo);
```

Comment:
How about adding `continue` here (we don't process this segment if we couldn't get the `directory`)?